app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java

```diff
@@ -0,0 +1,310 @@
+            showErrorDialog(Localization.get("print.not.supported"));
+        }
+
+        Bundle data = getIntent().getExtras();
+        //Check to make sure key-value data has been passed with the intent
+        if (data == null) {
+            showErrorDialog(Localization.get("no.print.data"));
+        }
+
+        this.outputPath = CommCareApplication._().getTempFilePath() + ".html";
+
+        //Check if a doc location is coming in from the Intent
+        //Will return a reference of format jr://... if it has been set
+        String path = data.getString("cc:print_template_reference");
+        if (path != null) {
+            try {
+                path = ReferenceManager._().DeriveReference(path).getLocalURI();
+                preparePrintDoc(path);
+            } catch (InvalidReferenceException e) {
+                showErrorDialog(Localization.get("template.invalid"));
+            }
+        } else {
+            //Try to use the document location that was set in Settings menu
+            SharedPreferences prefs = CommCareApplication._().getCurrentApp().getAppPreferences();
+            path = prefs.getString(CommCarePreferences.PRINT_DOC_LOCATION, "");
+            if ("".equals(path)) {
+                showErrorDialog(Localization.get("missing.template.file"));
+            } else {
+                preparePrintDoc(path);
+            }
+        }
+    }
+
+    private void preparePrintDoc(String inputPath) {
+        generateJobName(inputPath);
+        String extension = TemplatePrinterUtils.getExtension(inputPath);
+        File templateFile = new File(inputPath);
+        if (extension.equalsIgnoreCase("html") && templateFile.exists()) {
+            new TemplatePrinterTask(
+                    templateFile,
+                    outputPath,
+                    getIntent().getExtras(),
+                    this
+            ).execute();
+        } else {
+            showErrorDialog(Localization.get("template.invalid"));
+        }
+    }
+
+    /**
+     * Generate a unique name for this print job, using the name of the template file and the date
+     *
+     * @param templateFilename the path to the given template file
+     */
+    private void generateJobName(String templateFilename) {
+        String inputWithoutExtension = templateFilename.substring(0,
+                templateFilename.lastIndexOf('.'));
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+        String dateString = sdf.format(new Date());
+        jobName = inputWithoutExtension + "_" + dateString;
+    }
+
+    /**
+     * Called when TemplatePrinterTask finishes, with a result code indicating what happened. If
+     * a .html file of the filled-out template has been created and saved successfully, proceeds
+     * with printing. Otherwise, displays the appropriate error message.
+     */
+    @Override
+    public void onFinished(int result, String problemString) {
```

Comment:
Consider renaming to templatePopulationFinished or something that makes it clear what is finishing so that you don't need to rely on the comments to provide the context.